core/src/main/scala/medeia/generic/HlistEncoderInstances.scala

```diff
@@ -18,11 +17,10 @@ trait HlistEncoderInstances {
       options: GenericDerivationOptions[Base] = GenericDerivationOptions[Base]()
   ): ShapelessEncoder[Base, FieldType[K, H] :: T] = {
     val fieldName: String = options.transformKeys(witness.value.name)
-    hlist =>
+    (hlist, doc) =>
       {
         val head = hEncoder.value.encode(hlist.head)
-        val tail: BsonDocument = tEncoder.encode(hlist.tail)
-        tail.append(fieldName, head)
```

Comment:
So the problem is that `BsonDocument` is using a `LinkedHashMap` and here we `append` the `head` to the `tail`, so it's always putting it to the right.  

One solution would probably be to construct the bson document differently, by creating an empty `BsonDocument`, then putting the head followed by the tail.

That would make it less performant, but not sure if that's really a problem

Reply:
The api of BsonDocument does not have a good api for putting the tail. 
```public void putAll(final Map<? extends String, ? extends BsonValue> m)```

That would mean something like
``` scala
val doc1 = ???
val doc2 = ???
doc1.putAll(doc2.entrySet().asScala.map(x => (x.getKey -> x.getValue)).toMap.asJava)
```

Reply:
For example like this: ea521172cac75aa340aa0415ae4487fa4a544aa2

Reply:
We could also switch the target encoding away from `BsonDocument`, e.g. into a `SortedMap` and then do a final conversion step at the very end :thinking:

Reply:
The mutable `BsonDocument` is a bit ugly anyway :monocle_face:

Reply:
Ah nice, did not know the ChainingSyntax, that's good to know. But I would say performance is of importance.

Yeah most of these problems come from BsonDocument. The Change to SortedMap or something similar would be nice but would have to take place through the whole libary. Should do that in 2.0 ;)

Reply:
Hmm one other idea could be to "cheat-in"  a final post-processing step, e.g. by adding a `map` method to the encoder interface and somehow add a call to that before returning the final derived encoder.

In that step we could reverse the BsonDocument order, and do it only once. 

Not sure :shrug:  I guess passing in an accumulator is the most-performant way, but I somehow dislike deviating too much from the standard derivation shape.  Those problems might re-surface e.g. when migrating to Scala 3 and the built-in default derivation provided there...

Reply:
The problem is that in the current implementation both coproduct and hlist encoders share the same interface and coproduct uses BsonDocumentEncoder under the hood, which makes it hard to solve this bug without changing a lot around it or without hacky solutions. I believe that this shows that the current design is not optimal. I think I will have a look at rewriting part of medeia when mongo releases it's library for scala3 and see what can be made better.

Reply:
I would be okay with the accumulator solution for now, if order is important :smirk: :shrug: :grin: